File: code/python/剑指offer59-1.py
import queue
class Solution:
    def maxSlidingWindow(self, nums: List[int], k: int) -> List[int]:
        deque = queue.deque()
        res = []
        for i in range(k):
            num = nums[i]
            while deque and deque[-1]<num:
                deque.pop()
            deque.append(num)
        if deque: res.append(deque[0])
        for i in range(k,len(nums)):
            num = nums[i]
            out = nums[i-k]
            if out==deque[0]:
                deque.popleft()
            while deque and deque[-1]<num:
                deque.pop()
            deque.append(num)
            res.append(deque[0])
        return res

File: code/java/剑指offer59-1.java
class Solution {
    public int[] maxSlidingWindow(int[] nums, int k) {
        int[] result = new int[nums.length-k+1];
        if(nums.length==0) return new int[0];
        Deque<Integer> deque = new LinkedList<>();
        for(int i=0;i<k;i++){
            while(!deque.isEmpty()&&deque.peekLast()<nums[i]) deque.pollLast();
            deque.offerLast(nums[i]);
        }
        result[0] = deque.peekFirst();
        for(int i=k;i<nums.length;i++){
            if(nums[i-k]==deque.peekFirst()) deque.pollFirst();
            while(!deque.isEmpty()&&deque.peekLast()<nums[i]) deque.pollLast();
            deque.offerLast(nums[i]);
            result[i-k+1] = deque.peekFirst();
        }
        return result;
    }
}